hw/ip/otbn/dv/rig/rig/gens/misaligned_load_store.py

```diff
@@ -0,0 +1,301 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import random
+from typing import Optional
+
+from shared.operand import ImmOperandType, RegOperandType, OptionOperandType
+from shared.insn_yaml import Insn, InsnsFile
+
+from ..config import Config
+from ..model import Model
+from ..program import ProgInsn, Program
+
+from ..snippet import ProgSnippet
+from ..snippet_gen import GenCont, GenRet, SnippetGen
+
+
+class MisalignedLoadStore(SnippetGen):
+    '''A snippet generator that generates random load/store instructions
+
+    Generated instructions are misaligned, therefore they end the
+    program.
+
+    '''
+
+    ends_program = True
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__()
+
+        self.insns = []
+        self.weights = []
+
+        self.sw = self._get_named_insn(insns_file, 'sw')
+        self.lw = self._get_named_insn(insns_file, 'lw')
+        self.bnlid = self._get_named_insn(insns_file, 'bn.lid')
+        self.bnsid = self._get_named_insn(insns_file, 'bn.sid')
+
+        for insn in [self.sw, self.bnsid, self.lw, self.bnlid]:
+            weight = cfg.insn_weights.get(insn.mnemonic)
+            if weight > 0:
+                self.weights.append(weight)
+                self.insns.append(insn)
+
+        # Check that at least one instruction has a positive weight
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        weights = self.weights
+        prog_insn = None
+        while prog_insn is None:
+            idx = random.choices(range(len(self.insns)), weights=weights)[0]
+            # At least one weight should be positive. This is guaranteed so
+            # long as fill_insn doesn't fail for absolutely everything. We know
+            # that doesn't happen, because we have some instructions (such as
+            # addi) where it will not fail.
+            assert weights[idx] > 0
```

Comment:
This comment doesn't make sense here (I think it's copied from `StraightLineInsn`). I think it's also not actually true: I don't see why `fill_insn` can't fail for all 4 instructions. However, if all the weights are zero then `weights[idx]` will be zero and you can just return `None` here to give up cleanly.